qmp/rpc/rpc.go

```diff
@@ -0,0 +1,674 @@
+		// Why this is needed remains a mystery.
+		Padding: [3]byte{0x1, 0x0, 0x0},
+		Name:    "qemu:///system",
+		Flags:   0,
+	}
+
+	buf, err := encode(&payload)
+	if err != nil {
+		return err
+	}
+
+	resp, err := rpc.request(procConnectOpen, programRemote, &buf)
+	if err != nil {
+		return err
+	}
+
+	r := <-resp
+	if r.Status != StatusOK {
+		return decodeError(r.Payload)
+	}
+
+	return nil
+}
+
+// Disconnect shuts down communication with the libvirt server.
+// The underlying net.Conn is not closed.
+func (rpc *RPC) Disconnect() error {
+	// close event streams
+	for id := range rpc.events {
+		if err := rpc.removeStream(id); err != nil {
+			return err
+		}
+	}
+
+	// inform libvirt we're done
+	resp, err := rpc.request(procConnectClose, programRemote, nil)
+	if err != nil {
+		return err
+	}
+
+	r := <-resp
+	if r.Status != StatusOK {
+		return decodeError(r.Payload)
+	}
+
+	return nil
+}
+
+// Events streams QEMU QMP Events.
+// If a problem is encountered setting up the event monitor connection
+// an error will be returned. Errors encountered during streaming will
+// cause the returned event channel to be closed.
+func (rpc *RPC) Events() (<-chan qmp.Event, error) {
+	d, err := rpc.lookup(rpc.Domain)
+	if err != nil {
+		return nil, err
+	}
+
+	payload := struct {
+		Padding [4]byte
+		Domain  domain
+		Event   [2]byte
+		Flags   [2]byte
+	}{
+		Padding: [4]byte{0x0, 0x0, 0x1, 0x0},
+		Domain:  *d,
+		Event:   [2]byte{0x0, 0x0},
+		Flags:   [2]byte{0x0, 0x0},
+	}
+
+	buf, err := encode(&payload)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := rpc.request(qemuConnectDomainMonitorEventRegister, programQEMU, &buf)
+	if err != nil {
+		return nil, err
+	}
+
+	res := <-resp
+	dec := xdr.NewDecoder(bytes.NewReader(res.Payload))
+
+	cbID, _, err := dec.DecodeUint()
+	if err != nil {
+		return nil, err
+	}
+
+	c := make(chan qmp.Event)
+	go func() {
+		stream := make(chan *event)
+		rpc.addStream(cbID, stream)
+
+		// process events
+		for e := range stream {
+			qe, err := qmpEvent(e)
+			if err != nil {
+				close(c)
+				break
+			}
+
+			c <- *qe
+		}
+	}()
+
+	return c, nil
+}
+
+// Run executes the given QAPI command against a domain's QEMU instance.
+// For a list of available QAPI commands, see:
+//	http://git.qemu.org/?p=qemu.git;a=blob;f=qapi-schema.json;hb=HEAD
+func (rpc *RPC) Run(cmd []byte) ([]byte, error) {
+	d, err := rpc.lookup(rpc.Domain)
+	if err != nil {
+		return nil, err
+	}
+
+	payload := struct {
+		Domain  domain
+		Command []byte
+		Flags   uint32
+	}{
+		Domain:  *d,
+		Command: cmd,
+		Flags:   0,
+	}
+
+	buf, err := encode(&payload)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := rpc.request(qemuDomainMonitor, programQEMU, &buf)
+	if err != nil {
+		return nil, err
+	}
+
+	res := <-resp
+	if res.Status != StatusOK {
+		return nil, decodeError(res.Payload)
+	}
+
+	r := bytes.NewReader(res.Payload)
+	dec := xdr.NewDecoder(r)
+	data, _, err := dec.DecodeFixedOpaque(int32(r.Len()))
```

Comment:
Need to handle error here in case `data` is nil, so we don't reslice a nil pointer.